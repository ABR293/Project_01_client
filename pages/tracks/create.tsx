import { Button, Grid, TextField } from "@mui/material";
import axios from "axios";
import { useRouter } from "next/router";
import React, { useState } from "react";
import Api from "../../api";
import FileUploader from "../../components/FileUploader";
import StepWrapper from "../../components/StepWrapper";
import { useInput } from "../../hooks/useInput";
import MainLayout from "../../layouts/MainLayout";
import styles from "../../styles/create.module.sass";

const Create = () => {
  const [activeStep, setActiveStep] = useState(0);
  const [picture, setPicture] = useState(null);
  const [audio, setAudio] = useState(null);
  const [loading, setLoading] = useState(false);
  const router = useRouter();

  const name = useInput("");
  const artist = useInput("");
  const text = useInput("");

  const next = async () => {
    if (activeStep !== 2) {
      setActiveStep((prew) => prew + 1);
    } else {
      const formData = new FormData();
      formData.append("name", name.value);
      formData.append("artist", artist.value);
      formData.append("text", text.value);
      picture && formData.append("picture", picture);
      audio && formData.append("audio", audio);
      setLoading(true);
      try {
        axios.post("http://localhost:4000/tracks", formData);
        await Api.createTrack(formData);
        setLoading(false);
        router.push("/tracks");
      } catch (err) {
        console.log(err);
        setLoading(false);
      }
    }
  };
  const back = () => {
    setActiveStep((prew) => prew - 1);
  };
  return (
    <MainLayout>
      <StepWrapper activeStep={activeStep} data-testid="create">
        {activeStep === 0 && (
          <Grid container flexDirection="column" className={styles.container}>
            <TextField
              {...name}
              className={styles.input}
              label={"Название трека"}
            />
            <TextField
              {...artist}
              className={styles.input}
              label={"Исполнитель"}
            />
            <TextField
              {...text}
              className={styles.input}
              label={"Слова песни"}
            />
          </Grid>
        )}
        {activeStep === 1 && (
          <div>
            <FileUploader setFile={setPicture} accept="image/*">
              <Button>Загрузите обложку</Button>
            </FileUploader>
          </div>
        )}
        {activeStep === 2 && (
          <div>
            <FileUploader setFile={setAudio} accept="audio/*">
              <Button>Загрузите аудио</Button>
            </FileUploader>
          </div>
        )}
      </StepWrapper>
      <Grid container justifyContent="space-between">
        <Button disabled={activeStep === 0} onClick={back}>
          Назад
        </Button>
        <Button disabled={activeStep === 3} onClick={next}>
          Вперед
        </Button>
      </Grid>
    </MainLayout>
  );
};

export default Create;
