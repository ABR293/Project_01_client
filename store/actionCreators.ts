

import {playerActions} from './playerSlicer'
import {trackActions} from './tracksSlicer'

const actionCreators = {
    ...playerActions,
    ...trackActions,
}

export default actionCreators