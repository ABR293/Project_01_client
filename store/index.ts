import {createWrapper, Context, MakeStore, HYDRATE} from 'next-redux-wrapper';
import { Store } from 'redux'
import {configureStore} from '@reduxjs/toolkit'
import createSagaMiddleware, {Task} from 'redux-saga'
import { combineReducers } from '@reduxjs/toolkit'
import playerReducer  from './playerSlicer';
import tracksReducer  from './tracksSlicer';
import rootSaga from './sagas';

 const rootReducer = combineReducers({
    player: playerReducer,
    tracks: tracksReducer
})

export type RootState = ReturnType<typeof rootReducer>


export interface SagaStore extends Store {
  sagaTask?: Task;
}

export const reducer = (state:any, action: any) => {
  if (action.type === HYDRATE) {
    const nextState = {
      ...state, // use previous state
      ...action.payload, // apply delta from hydration
    };
    if (state.count) nextState.count = state.count; // preserve count value on client side navigation
    return nextState;
  } else {
    return rootReducer(state, action);
  }
};

let sagaMiddleware = createSagaMiddleware()

export const store = configureStore({
    reducer,
    middleware: [sagaMiddleware],
})

const makeStore = (context: Context) => {
  (store as SagaStore).sagaTask = sagaMiddleware.run(rootSaga);
  return store
}

export const wrapper = createWrapper<Store<RootState>>(makeStore, {debug: true}); 

// sagaMiddleware.run(rootSaga)


//  middleware: (getDefaultMiddleware) =>
//     getDefaultMiddleware().concat(middleware),

