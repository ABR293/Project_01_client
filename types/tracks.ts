export interface IComment {
  _id: string;
  userName: string;
  text: string;
  likes: number;
  disLikes: number;
}

export type CreateComment = {
  userName: string;
  text: string;
  trackId: string;
};

export interface ITrack {
  _id: string;
  name: string;
  artist: string;
  text: string;
  listens: number;
  picture: string;
  audio: string;
  comments: IComment[];
}

export interface ITrackList {
  tracks: ITrack[];
}
