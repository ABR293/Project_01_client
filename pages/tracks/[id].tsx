import { Box } from '@mui/system'
import React from 'react'
import { IComment, ITrack } from '../../types/tracks'
import { useRouter } from 'next/router'
import MainLayout from '../../layouts/MainLayout'
import { Button, Grid, TextField } from '@mui/material';

interface ITrackPageProps {
    track: ITrack;
}
const TrackPage:React.FC = () => {
    const track =  {_id: '1', name: 'Track1', artist: 'Rammstein', text:'Sonne', listens: 4, picture: 'http://localhost:4000/image/459d481c-9f25-43eb-8b78-6fd805356da0.jpeg', audio: 'http://localhost:4000/audio/a01cbead-da79-4051-a90c-c70a8fd5fb8b.mp3', comments:[]};
    const router = useRouter()
    return (
        <MainLayout>
            <Button 
                onClick={() => router.push('/tracks')}
                variant='outlined'
                >
                К списку
            </Button>
            <Grid container >
                <img src={track.picture} width={200} height={200}/>
                <div>
                    <h1>{track.name}</h1>
                    <h1>{track.artist}</h1>
                    <h1>{track.listens}</h1>
                </div>
            </Grid>
            <h1>Текст песни </h1>
            <p>{track.text}</p>
            <Grid container>
                <TextField 
                    label='Ваше имя'
                    fullWidth
                />
                <TextField 
                    label='Комментарий'
                    fullWidth
                    multiline
                    rows={4}
                />
                <Button>Отправить</Button>
            </Grid>
            {/* <div>
                {track.comments.map(comment =>
                    <div key={comment._id}>
                        <h4>Автор - {comment.username}</h4>
                        <p>Комментарий - {comment.text}</p>
                    </div>
                )}
            </div> */}
        </MainLayout>
        // <>TrackPage</>
        // <Card className={styles.track} onClick={() => router.push('/tracks/'+ track._id)}>
        //     {/* {track.name} */}
        //     <IconButton>
        //         { active
        //             ? <Pause/>
        //             : <PlayArrow/>
        //         }
        //     </IconButton>
        //     <img width={70} height={70} src={track.picture} alt='нет изображения'/>
        //     <Box p={1}>
        //         <Grid direction='column'>
        //             <h3>{track.name}</h3>
        //             <h4>{track.artist}</h4>
        //         </Grid>
        //     </Box>
        //     <IconButton>
        //         <Delete />
        //     </IconButton>
            

        // </Card>
    )
}

export default TrackPage