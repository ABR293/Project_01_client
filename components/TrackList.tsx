import { Grid } from '@mui/material'
import { Box } from '@mui/system'
import React from 'react'
import { ITrack } from '../types/tracks'
import TrackItem from './TrackItem'

interface ITrackListProps {
    tracks: ITrack[]
}
const TrackList:React.FC<ITrackListProps> = ({tracks}) => {
    return (
        <Grid container direction='column'>
            <Box>
                {tracks.map(track => 
                    <TrackItem key={track._id} track={track}/>
                )}
            </Box>
        </Grid>
    )
}

export default TrackList