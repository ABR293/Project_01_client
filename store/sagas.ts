import { call, put, takeEvery, takeLatest } from 'redux-saga/effects'
import actionCreators from './actionCreators'
import axios, { Axios, AxiosResponse } from 'axios'
import { ITrack } from '../types/tracks';
import { ServerResponse } from 'http';
import Api from '../api';

function* workerFetchTracs({payload = ''}) {
   try {
      let data:ITrack[] = []
      if (payload !== '') {
         data = yield call(() => Api.searchTracks(payload));
      } else {
         data = yield call(() => Api.getAllTracks() );
      }
      yield put(actionCreators.fetchTracksSuccess(data));
   } catch (err) {
      yield put(actionCreators.fetchTracksError('Произошла ошибка при загрузке треков')); 
   }
}

function* tracksSaga() {
   // yield takeEvery('tracks/fetchTracks', workerFetchTracs)
   yield takeLatest('getTracks', workerFetchTracs)
}

export default  function* rootSaga() {
   yield tracksSaga()
}

