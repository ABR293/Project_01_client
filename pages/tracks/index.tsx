import { Button, Card, Grid } from '@mui/material';
import { Box } from '@mui/system';
import { useRouter } from 'next/router';
import React, { useTransition } from 'react'
import TrackList from '../../components/TrackList';
import { useTypedSelector } from '../../hooks/useTypedSelector';
import MainLayout from '../../layouts/MainLayout';
import { wrapper } from '../../store';
import { ITrack } from '../../types/tracks';
import actionCreators from '../../store/actionCreators';
// import {store} from '../../store';
import {END} from 'redux-saga';
import { GetServerSideProps } from 'next';

// import { Store } from 'redux'

const Index = () => {
    const router = useRouter();

    const state = useTypedSelector(state => state)
    const {error, tracks} = state.tracks;

    if(error){
        return <MainLayout><h1>{error}</h1></MainLayout>
    }

    return (
        <MainLayout>
            <Grid container justifyContent='center'>
                <Card style={{width: '80%'}}>
                    <Box p={2}>
                        <Grid container justifyContent='space-between'>
                            <h1>Список треков</h1>
                            <Button onClick={() => {router.push('/tracks/create')}}>Загрузить</Button>
                        </Grid>
                    </Box>
                    
                </Card>
            </Grid>
            <TrackList tracks={tracks} /> 
        </MainLayout>

    )
}

export default Index

export const getServerSideProps:GetServerSideProps = wrapper.getServerSideProps((store) => async () => {
     // regular stuff
  store.dispatch(actionCreators.fetchTracks());
  // end the saga
  store.dispatch(END);
  await store.sagaTask.toPromise();
})
