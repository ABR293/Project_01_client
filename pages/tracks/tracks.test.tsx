import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "../../store/index";
import Tracks from "./index";
import Create from "./create";
import TrackList from "../../components/TrackList";
import { ITrack } from "../../types/tracks";

jest.mock("next/router", () => ({
  useRouter() {
    return {
      route: "/",
      pathname: "",
      query: "",
      asPath: "",
      push: jest.fn(),
    };
  },
}));

const tracks: ITrack[] = [
  {
    _id: "23921389",
    name: "Du hast",
    artist: "Rammstein",
    text: "Du! Du Hast! Du Hast Mich!",
    listens: 5,
    picture: "test/pic",
    audio: "test/audio",
    comments: [],
  },
  {
    _id: "239213f9",
    name: "Wind of change",
    artist: "Scorpions",
    text: "I follow the Moskva \n Down to Gorky Park \n Listening to the wind of change",
    listens: 5,
    picture: "test/pic",
    audio: "test/audio",
    comments: [],
  },
];

describe("Render Trsacks", () => {
  it("Tracks", () => {
    render(
      <Provider store={store}>
        <Tracks />
      </Provider>
    );

    expect(screen.queryByTestId("tracks")).toMatchSnapshot();
    screen.debug;
  });
  it("Create", () => {
    render(
      <Provider store={store}>
        <Create />
      </Provider>
    );

    expect(screen.queryByTestId("create")).toMatchSnapshot();
    screen.debug;
  });

  it("render TrackList", () => {
    render(
      <Provider store={store}>
        <TrackList tracks={tracks} />
      </Provider>
    );
    expect(screen.getAllByTestId("trackItem")).toMatchSnapshot();
    expect(screen.getByTestId("trackList")).toMatchSnapshot();
    // screen.debug
  });
});
