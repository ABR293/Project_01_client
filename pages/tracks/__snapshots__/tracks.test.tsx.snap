// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Trsacks Create 1`] = `null`;

exports[`Render Trsacks Tracks 1`] = `null`;

exports[`Render Trsacks render TrackList 1`] = `
[
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root track css-bhp9pd-MuiPaper-root-MuiCard-root"
    data-testid="trackItem"
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="PlayArrowIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M8 5v14l11-7z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <img
      alt=""
      height="60"
      src="http://localhost:4000/test/pic"
      width="60"
    />
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-1vpwcmr-MuiGrid-root"
      >
        <h3>
          Du hast
        </h3>
        <h5>
          Rammstein
        </h5>
      </div>
    </div>
    <div
      class="MuiBox-root css-1qapnz0"
    />
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="AddIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>,
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root track css-bhp9pd-MuiPaper-root-MuiCard-root"
    data-testid="trackItem"
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="PlayArrowIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M8 5v14l11-7z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <img
      alt=""
      height="60"
      src="http://localhost:4000/test/pic"
      width="60"
    />
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-1vpwcmr-MuiGrid-root"
      >
        <h3>
          Wind of change
        </h3>
        <h5>
          Scorpions
        </h5>
      </div>
    </div>
    <div
      class="MuiBox-root css-1qapnz0"
    />
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="AddIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>,
]
`;

exports[`Render Trsacks render TrackList 2`] = `
<div
  class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-1vpwcmr-MuiGrid-root"
  data-testid="trackList"
>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root track css-bhp9pd-MuiPaper-root-MuiCard-root"
      data-testid="trackItem"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="PlayArrowIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M8 5v14l11-7z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <img
        alt=""
        height="60"
        src="http://localhost:4000/test/pic"
        width="60"
      />
      <div
        class="MuiBox-root css-0"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-1vpwcmr-MuiGrid-root"
        >
          <h3>
            Du hast
          </h3>
          <h5>
            Rammstein
          </h5>
        </div>
      </div>
      <div
        class="MuiBox-root css-1qapnz0"
      />
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="AddIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root track css-bhp9pd-MuiPaper-root-MuiCard-root"
      data-testid="trackItem"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="PlayArrowIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M8 5v14l11-7z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <img
        alt=""
        height="60"
        src="http://localhost:4000/test/pic"
        width="60"
      />
      <div
        class="MuiBox-root css-0"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-1vpwcmr-MuiGrid-root"
        >
          <h3>
            Wind of change
          </h3>
          <h5>
            Scorpions
          </h5>
        </div>
      </div>
      <div
        class="MuiBox-root css-1qapnz0"
      />
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="AddIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
